File: cloud/api/src/Modules/Sync/Contract/ChangeDetectionEngine.php
<?php

declare(strict_types=1);

namespace MizaCloud\Modules\Sync\Contract;

/**
 * Field Change Detection Engine — Field Change Detection Specification v1.0.
 *
 * Pure comparison: base snapshot vs next snapshot → changed_fields map.
 * Clients and cloud must produce identical results for the same inputs + dictionary.
 */
final class ChangeDetectionEngine
{
    public function __construct(
        private readonly FieldMetadataProvider $metadata,
    ) {
    }

    /**
     * Detect changed patchable fields between base and next entity snapshots.
     *
     * @param array<string, mixed> $baseSnapshot field => value (previous accepted state)
     * @param array<string, mixed> $nextSnapshot field => value (intended new state)
     * @return array<string, mixed> changed_fields map (field => next normalized value)
     */
    public function detect(string $entityType, array $baseSnapshot, array $nextSnapshot): array
    {
        $entity = $this->metadata->entity($entityType);
        $changed = [];

        foreach ($entity->patchableFields() as $field) {
            $baseRaw = array_key_exists($field->name, $baseSnapshot)
                ? $baseSnapshot[$field->name]
                : $field->defaultValue;
            $nextRaw = array_key_exists($field->name, $nextSnapshot)
                ? $nextSnapshot[$field->name]
                : $field->defaultValue;

            $baseNorm = $this->normalize($field, $baseRaw);
            $nextNorm = $this->normalize($field, $nextRaw);

            if (!$this->valuesEqual($field, $baseNorm, $nextNorm)) {
                $changed[$field->name] = $nextNorm;
            }
        }

        return $changed;
    }

    /**
     * Normalize a single field value for comparison / patch emission.
     */
    public function normalize(FieldSpec $field, mixed $value): mixed
    {
        if ($value === null) {
            if (!$field->nullable && $field->defaultValue !== null) {
                return $this->normalize($field, $field->defaultValue);
            }

            return null;
        }

        if ($field->emptyAsNull && is_string($value) && trim($value) === '') {
            return null;
        }

        return match ($field->comparisonRule) {
            'string_default' => $this->normalizeString($value),
            'decimal_scale' => $this->normalizeDecimal($value, $field->decimalScale ?? 4),
            'boolean_default' => $this->normalizeBoolean($value),
            'integer_default' => $this->normalizeInteger($value),
            'uuid_canonical' => $this->normalizeUuid($value),
            'datetime_iso8601' => $this->normalizeDateTime($value),
            'image_url_special' => $this->normalizeImageUrl($value, $field->emptyAsNull),
            default => throw new \InvalidArgumentException(
                'Unsupported comparison_rule "' . $field->comparisonRule . '" for field ' . $field->name
            ),
        };
    }

    public function valuesEqual(FieldSpec $field, mixed $a, mixed $b): bool
    {
        if ($a === null && $b === null) {
            return true;
        }
        if ($a === null || $b === null) {
            return false;
        }

        return match ($field->comparisonRule) {
            'decimal_scale' => (string) $a === (string) $b,
            'boolean_default' => (bool) $a === (bool) $b,
            'integer_default' => (int) $a === (int) $b,
            default => $a === $b,
        };
    }

    private function normalizeString(mixed $value): string
    {
        if (is_bool($value) || is_array($value) || is_object($value)) {
            throw new \InvalidArgumentException('string field cannot accept complex/bool value');
        }

        return (string) $value;
    }

    private function normalizeDecimal(mixed $value, int $scale): string
    {
        if (is_bool($value) || is_array($value) || is_object($value) || $value === null) {
            throw new \InvalidArgumentException('decimal field requires numeric value');
        }
        if (is_string($value) && trim($value) === '') {
            throw new \InvalidArgumentException('decimal field cannot be empty string');
        }

        if (!is_numeric($value)) {
            throw new \InvalidArgumentException('decimal field requires numeric value');
        }

        return number_format((float) $value, $scale, '.', '');
    }

    private function normalizeBoolean(mixed $value): bool
    {
        if (is_bool($value)) {
            return $value;
        }
        if (is_int($value) || is_float($value)) {
            if ($value === 1 || $value === 1.0) {
                return true;
            }
            if ($value === 0 || $value === 0.0) {
                return false;
            }
            throw new \InvalidArgumentException('boolean field numeric must be 0 or 1');
        }
        if (is_string($value)) {
            $v = strtolower(trim($value));
            if (in_array($v, ['1', 'true', 'yes'], true)) {
                return true;
            }
            if (in_array($v, ['0', 'false', 'no', ''], true)) {
                return false;
            }
        }

        throw new \InvalidArgumentException('boolean field has invalid value');
    }

    private function normalizeInteger(mixed $value): int
    {
        if (is_bool($value) || is_array($value) || is_object($value)) {
            throw new \InvalidArgumentException('integer field has invalid value');
        }
        if (is_int($value)) {
            return $value;
        }
        if (is_float($value)) {
            if (floor($value) !== $value) {
                throw new \InvalidArgumentException('integer field cannot be fractional');
            }

            return (int) $value;
        }
        if (is_string($value)) {
            $v = trim($value);
            if ($v === '' || !preg_match('/^-?\d+$/', $v)) {
                throw new \InvalidArgumentException('integer field has invalid value');
            }

            return (int) $v;
        }

        throw new \InvalidArgumentException('integer field has invalid value');
    }

    private function normalizeUuid(mixed $value): ?string
    {
        if ($value === null) {
            return null;
        }
        if (!is_string($value)) {
            throw new \InvalidArgumentException('uuid field must be string or null');
        }
        $v = strtolower(trim($value));
        if ($v === '') {
            return null;
        }
        if (!preg_match('/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/', $v)) {
            throw new \InvalidArgumentException('uuid field has invalid format');
        }

        return $v;
    }

    private function normalizeDateTime(mixed $value): ?string
    {
        if ($value === null) {
            return null;
        }
        if (!is_string($value) || trim($value) === '') {
            throw new \InvalidArgumentException('datetime field must be non-empty ISO-8601 string or null');
        }
        $ts = strtotime($value);
        if ($ts === false) {
            throw new \InvalidArgumentException('datetime field has invalid value');
        }

        return gmdate('Y-m-d\TH:i:s\Z', $ts);
    }

    private function normalizeImageUrl(mixed $value, bool $emptyAsNull): ?string
    {
        if ($value === null) {
            return null;
        }
        if (!is_string($value)) {
            throw new \InvalidArgumentException('image_url field must be string or null');
        }
        $v = trim($value);
        if ($v === '') {
            return $emptyAsNull ? null : '';
        }

        return $v;
    }
}

File: apps/mizapos_desktop/lib/services/cloud/sync/contract/change_detection_engine.dart
import 'package:mizapos_desktop/services/cloud/sync/contract/field_spec.dart';
import 'package:mizapos_desktop/services/cloud/sync/contract/product_field_dictionary.dart';

/// Field Change Detection Engine — Spec v1.0 (client mirror).
///
/// Defaults to Products patchable fields; pass [fields] for Partners / others.
abstract final class ChangeDetectionEngine {
  /// Returns changed_fields map (field → normalized next value).
  static Map<String, Object?> detect({
    required Map<String, Object?> baseSnapshot,
    required Map<String, Object?> nextSnapshot,
    List<FieldSpec>? fields,
  }) {
    final changed = <String, Object?>{};
    final specs = fields ?? ProductFieldDictionary.patchableFields;
    for (final field in specs) {
      final baseRaw = baseSnapshot.containsKey(field.name)
          ? baseSnapshot[field.name]
          : field.defaultValue;
      final nextRaw = nextSnapshot.containsKey(field.name)
          ? nextSnapshot[field.name]
          : field.defaultValue;
      final baseNorm = normalize(field, baseRaw);
      final nextNorm = normalize(field, nextRaw);
      if (!valuesEqual(field, baseNorm, nextNorm)) {
        changed[field.name] = nextNorm;
      }
    }
    return changed;
  }

  static Object? normalize(FieldSpec field, Object? value) {
    if (value == null) {
      if (!field.nullable && field.defaultValue != null) {
        return normalize(field, field.defaultValue);
      }
      return null;
    }
    if (field.emptyAsNull && value is String && value.trim().isEmpty) {
      return null;
    }

    switch (field.comparisonRule) {
      case 'string_default':
        return value.toString();
      case 'decimal_scale':
        return _normalizeDecimal(value, field.decimalScale ?? 4);
      case 'boolean_default':
        return _normalizeBool(value);
      case 'integer_default':
        return _normalizeInt(value);
      case 'uuid_canonical':
        return _normalizeUuid(value);
      case 'image_url_special':
        return _normalizeImageUrl(value, field.emptyAsNull);
      case 'datetime_iso8601':
        return _normalizeDateTime(value);
      default:
        throw ArgumentError(
          'Unsupported comparison_rule ${field.comparisonRule} for ${field.name}',
        );
    }
  }

  static bool valuesEqual(FieldSpec field, Object? a, Object? b) {
    if (a == null && b == null) return true;
    if (a == null || b == null) return false;
    switch (field.comparisonRule) {
      case 'decimal_scale':
        return a.toString() == b.toString();
      case 'boolean_default':
        return a == b;
      case 'integer_default':
        return a == b;
      default:
        return a == b;
    }
  }

  static String _normalizeDecimal(Object? value, int scale) {
    if (value is bool) {
      throw ArgumentError('decimal field requires numeric value');
    }
    final n = value is num ? value.toDouble() : double.tryParse(value.toString());
    if (n == null) {
      throw ArgumentError('decimal field requires numeric value');
    }
    return n.toStringAsFixed(scale);
  }

  static bool _normalizeBool(Object? value) {
    if (value is bool) return value;
    if (value is num) {
      if (value == 1) return true;
      if (value == 0) return false;
      throw ArgumentError('boolean field numeric must be 0 or 1');
    }
    final v = value.toString().trim().toLowerCase();
    if (v == '1' || v == 'true' || v == 'yes') return true;
    if (v == '0' || v == 'false' || v == 'no' || v.isEmpty) return false;
    throw ArgumentError('boolean field has invalid value');
  }

  static int _normalizeInt(Object? value) {
    if (value is int) return value;
    if (value is num) {
      if (value != value.roundToDouble()) {
        throw ArgumentError('integer field cannot be fractional');
      }
      return value.toInt();
    }
    final parsed = int.tryParse(value.toString().trim());
    if (parsed == null) {
      throw ArgumentError('integer field has invalid value');
    }
    return parsed;
  }

  static String? _normalizeUuid(Object? value) {
    if (value == null) return null;
    final v = value.toString().trim().toLowerCase();
    if (v.isEmpty) return null;
    final re = RegExp(
      r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$',
    );
    if (!re.hasMatch(v)) {
      throw ArgumentError('uuid field has invalid format');
    }
    return v;
  }

  static String? _normalizeImageUrl(Object? value, bool emptyAsNull) {
    if (value == null) return null;
    final v = value.toString().trim();
    if (v.isEmpty) return emptyAsNull ? null : '';
    return v;
  }

  static String? _normalizeDateTime(Object? value) {
    if (value == null) return null;
    final raw = value.toString().trim();
    if (raw.isEmpty) {
      throw ArgumentError('datetime field must be non-empty ISO-8601 string or null');
    }
    final dt = DateTime.tryParse(raw);
    if (dt == null) {
      throw ArgumentError('datetime field has invalid value');
    }
    return dt.toUtc().toIso8601String().replaceFirst(RegExp(r'\.\d+'), '');
  }
}
